### PR TITLE
Support AES-256 encryption with key-rotation

### PR DESCRIPTION
**Pull request by hwupathum**
Related issue https://github.com/wso2/product-is/issues/21217

### CODE REVIEW

File: components/org.wso2.carbon.crypto.provider/src/test/java/org/wso2/carbon/crypto/provider/SymmetricKeyInternalCryptoProviderTest.java [hunk 1]
```diff
@@ -0,0 +1,111 @@
+package org.wso2.carbon.crypto.provider;
```
Comment (ashendes) [inline components/org.wso2.carbon.crypto.provider/src/test/java/org/wso2/carbon/crypto/provider/SymmetricKeyInternalCryptoProviderTest.java:18]: Add license

File: components/org.wso2.carbon.crypto.provider/src/test/java/org/wso2/carbon/crypto/provider/SymmetricKeyInternalCryptoProviderTest.java [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+        assertEquals(new String(Base64.getEncoder().encode(encryptedText)), CIPHER_TEXT);
+    }
+
+    @Test(description = "Test encryption using a custom secret.")
+    public void testEncryptUsingCustomSecret() throws CryptoException {
+
+        byte[] plainTextBytes = PLAIN_TEXT.getBytes();
+        byte[] encryptedText = symmetricKeyInternalCryptoProvider.encrypt(
+                plainTextBytes, AES_ALGORITHM, JCE_PROVIDER, false, OLD_SECRET);
+        // TODO: should return CIPHER_TEXT_FOR_OLD_SECRET
```
Comment (ashendes) [inline components/org.wso2.carbon.crypto.provider/src/test/java/org/wso2/carbon/crypto/provider/SymmetricKeyInternalCryptoProviderTest.java]:
```suggestion
        // TODO: Should return CIPHER_TEXT_FOR_OLD_SECRET. Update once SymmetricKeyInternalCryptoProvider behaviour is fixed.
```

File: components/org.wso2.carbon.crypto.provider/src/test/java/org/wso2/carbon/crypto/provider/SymmetricKeyInternalCryptoProviderTest.java [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+                encryptedText, AES_ALGORITHM, JCE_PROVIDER, OLD_SECRET);
+        assertEquals(new String(decryptedText), PLAIN_TEXT);
+    }
+
+    @Test(description = "Test decryption using an invalid custom secret.")
+    public void testDecryptUsingInvalidCustomSecret() throws CryptoException {
+
+        byte[] encryptedText = Base64.getDecoder().decode(CIPHER_TEXT);
+        byte[] decryptedText = symmetricKeyInternalCryptoProvider.decrypt(
+                encryptedText, AES_ALGORITHM, JCE_PROVIDER, OLD_SECRET);
```
Comment (ashendes) [inline components/org.wso2.carbon.crypto.provider/src/test/java/org/wso2/carbon/crypto/provider/SymmetricKeyInternalCryptoProviderTest.java]: Decryption with old secret works but encryption doesn't?

Comment (hwupathum) [reply components/org.wso2.carbon.crypto.provider/src/test/java/org/wso2/carbon/crypto/provider/SymmetricKeyInternalCryptoProviderTest.java]: Seems like the variable name is misleading. Will update

File: components/org.wso2.carbon.crypto.provider/src/test/java/org/wso2/carbon/crypto/provider/SymmetricKeyInternalCryptoProviderTest.java [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+        assertEquals(new String(Base64.getEncoder().encode(encryptedText)), CIPHER_TEXT);
+    }
+
+    @Test(description = "Test encryption using a custom secret.")
+    public void testEncryptUsingCustomSecret() throws CryptoException {
+
+        byte[] plainTextBytes = PLAIN_TEXT.getBytes();
+        byte[] encryptedText = symmetricKeyInternalCryptoProvider.encrypt(
+                plainTextBytes, AES_ALGORITHM, JCE_PROVIDER, false, OLD_SECRET);
+        // TODO: Should return CIPHER_TEXT_FOR_OLD_SECRET. Update once SymmetricKeyInternalCryptoProvider behaviour is fixed.
```
Comment (ashendes) [inline components/org.wso2.carbon.crypto.provider/src/test/java/org/wso2/carbon/crypto/provider/SymmetricKeyInternalCryptoProviderTest.java]: Line longer than 120 characters. Fix formatting in IDE.

File: components/org.wso2.carbon.crypto.provider/src/main/java/org/wso2/carbon/crypto/provider/internal/DefaultCryptoProviderComponent.java [hunk 5]
```diff
@@ -57,6 +56,8 @@
 public class DefaultCryptoProviderComponent {
 
     public static final String CRYPTO_SECRET_PROPERTY_PATH = "CryptoService.Secret";
+    public static final String CRYPTO_OLD_SECRET_PROPERTY_PATH = "CryptoService.OldSecret.Secret";
```
Comment (hwupathum) [inline components/org.wso2.carbon.crypto.provider/src/main/java/org/wso2/carbon/crypto/provider/internal/DefaultCryptoProviderComponent.java]: made private

File: components/org.wso2.carbon.crypto.api/src/main/java/org/wso2/carbon/crypto/api/CipherMetaDataHolder.java [hunk 6]
```diff
@@ -153,6 +191,10 @@ public void setIvAndOriginalCipherText(byte[] cipherTextBytes) {
                 gson.fromJson(cipherStr, CipherInitializationVectorHolder.class);
         setIv(cipherInitializationVectorHolder.getInitializationVector());
         setCipherText(cipherInitializationVectorHolder.getCipher());
+        String keyId = cipherInitializationVectorHolder.getKeyId();
+        if (StringUtils.isNotBlank(keyId)) {
```
Comment (madurangasiriwardena) [inline components/org.wso2.carbon.crypto.api/src/main/java/org/wso2/carbon/crypto/api/CipherMetaDataHolder.java:195]: Let's update the method comment to notify the new change.

File: components/org.wso2.carbon.crypto.provider/src/test/java/org/wso2/carbon/crypto/provider/SymmetricKeyInternalCryptoProviderTest.java [hunk 7]
```diff
@@ -0,0 +1,225 @@
...
+
+    @BeforeClass
+    public void init() {
+
+        cryptoProvider = new SymmetricKeyInternalCryptoProvider(SECRET);
+        cryptoProviderWithBackupSecret = new SymmetricKeyInternalCryptoProvider(SECRET, OLD_SECRET, true);
+    }
+
+    // Default tests.
+
```
Comment (madurangasiriwardena) [inline components/org.wso2.carbon.crypto.provider/src/test/java/org/wso2/carbon/crypto/provider/SymmetricKeyInternalCryptoProviderTest.java]: unnecessary new line

File: components/org.wso2.carbon.crypto.provider/src/test/java/org/wso2/carbon/crypto/provider/SymmetricKeyInternalCryptoProviderTest.java [hunk 8]
```diff
@@ -0,0 +1,225 @@
...
+        assertEquals(new String(Base64.getEncoder().encode(encryptedText)), CIPHER_TEXT);
+    }
+
+    @Test(description = "Test encryption using a old secret.")
+    public void testEncryptUsingOldSecret() throws CryptoException {
+
+        byte[] plainTextBytes = PLAIN_TEXT.getBytes();
+        byte[] encryptedText = cryptoProvider.encrypt(
+                plainTextBytes, AES_ALGORITHM, JCE_PROVIDER, false, OLD_SECRET);
+        // TODO: Should return CIPHER_TEXT_FOR_OLD_SECRET.
```
Comment (madurangasiriwardena) [inline components/org.wso2.carbon.crypto.provider/src/test/java/org/wso2/carbon/crypto/provider/SymmetricKeyInternalCryptoProviderTest.java]: Either fix this todo or create a git issue for this.

File: components/org.wso2.carbon.crypto.provider/src/test/java/org/wso2/carbon/crypto/provider/SymmetricKeyInternalCryptoProviderTest.java [hunk 9]
```diff
@@ -0,0 +1,225 @@
...
+    @Test(description = "Test decryption using an invalid old secret.")
+    public void testDecryptUsingInvalidOldSecret() throws CryptoException {
+
+        byte[] encryptedText = Base64.getDecoder().decode(CIPHER_TEXT);
+        byte[] decryptedText = cryptoProvider.decrypt(encryptedText, AES_ALGORITHM, JCE_PROVIDER, OLD_SECRET);
+        assertEquals(new String(decryptedText), PLAIN_TEXT);
+    }
+
+    // Default tests in AES-GCM mode.
+
```
Comment (madurangasiriwardena) [inline components/org.wso2.carbon.crypto.provider/src/test/java/org/wso2/carbon/crypto/provider/SymmetricKeyInternalCryptoProviderTest.java]: unnecessary new line.

File: components/org.wso2.carbon.crypto.provider/src/test/java/org/wso2/carbon/crypto/provider/SymmetricKeyInternalCryptoProviderTest.java (hunk 8 above)
Comment (hwupathum) [reply components/org.wso2.carbon.crypto.provider/src/test/java/org/wso2/carbon/crypto/provider/SymmetricKeyInternalCryptoProviderTest.java]: Fixed the todo

File: components/org.wso2.carbon.crypto.provider/src/main/java/org/wso2/carbon/crypto/provider/SymmetricKeyInternalCryptoProvider.java [hunk 10]
```diff
@@ -348,6 +328,61 @@ private byte[] encryptWithGCMMode(byte[] plaintext, String javaSecurityAPIProvid
         return cipherText;
     }
 
+    /**
+     * This method will encrypt a given plain text in AES/ECB/PKCS5Padding cipher transformation
+     *
+     * @param plaintext                     Plain text that need to be encrypted in this mode.
+     * @param javaSecurityAPIProvider       Crypto provider.
+     * @param returnSelfContainedCipherText Whether the ciphertext should be in self-contained  mode or not.
```
Comment (madurangasiriwardena) [inline components/org.wso2.carbon.crypto.provider/src/main/java/org/wso2/carbon/crypto/provider/SymmetricKeyInternalCryptoProvider.java:336]: Shouldn't we add the optional params as well to the doc comment?

File: components/org.wso2.carbon.crypto.provider/src/main/java/org/wso2/carbon/crypto/provider/SymmetricKeyInternalCryptoProvider.java [hunk 11]
```diff
@@ -348,6 +328,61 @@ private byte[] encryptWithGCMMode(byte[] plaintext, String javaSecurityAPIProvid
...
+            if (returnSelfContainedCipherText) {
+                return createSelfContainedCiphertextWithPlainAES(cipherText, DEFAULT_SYMMETRIC_CRYPTO_ALGORITHM);
+            }
+            return cipherText;
+        } catch (NoSuchAlgorithmException | NoSuchPaddingException | InvalidKeyException | BadPaddingException |
+                 NoSuchProviderException | IllegalBlockSizeException e) {
+
+            String errorMessage = String.format("An error occurred while encrypting using the algorithm : '%s'"
+                    , DEFAULT_SYMMETRIC_CRYPTO_ALGORITHM);
+            if(log.isDebugEnabled()){
```
Comment (madurangasiriwardena) [inline components/org.wso2.carbon.crypto.provider/src/main/java/org/wso2/carbon/crypto/provider/SymmetricKeyInternalCryptoProvider.java]: Formatting issue

**Review by madurangasiriwardena — APPROVED** [review]

**Pull request merged**